### PR TITLE
[WIP] Fusion and _intrinsic modules

### PR DESCRIPTION
**Pull request by z-a-f**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#23003 [WIP] Fusion and _intrinsic modules**&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D16199720/)

torch.quantization.fuse_module and torch.nn._intrinsic convRelu and LinearRelu

Fusion function to combine specific modules: (conv,bn) and  (conv,bn,relu).
In all cases, replace modules in place. The first module is replaced with the _intrinsic fused module and the remaining modules are replaced by nn.Identity.
Support both training and eval. For training, the modules are "fused" with a sequential container. This is to allow for further module swaps for quantization aware training.
Also add: torch.nn._intrinsic for convRelu and LinearRelu.

TODO: Add tests for _intrinsic modules.

Conv BN fusion code is based on DsKhudia's implementation

Differential Revision: [D16199720](https://our.internmc.facebook.com/intern/diff/D16199720/)

### CODE REVIEW

**Comment by vadimkantorov** [comment]
ReLU and Dropout can also be fused (inplace as well; since both are homogenous): https://gist.github.com/vadimkantorov/360ece06de4fd2641fa9ed1085f76d48

Conv+BN+Relu+Dropout are currently used in some recent speech recognition models ( https://github.com/NVIDIA/OpenSeq2Seq/blob/master/example_configs/speech2text/jasper10x5_LibriSpeech_nvgrad_masks.py )

**Comment by vadimkantorov** [comment]
Also, these fusion helpers are useful (and valid) even outside qunantization context

**Comment by jerryzh168** [comment]
@vadimkantorov the reason we put this in quantization package is because it is not going to work generally since we just combine modules, and don't have access to how it is used in forward. Full support for fusion will only come with graph mode(JIT).

File: torch/quantization/fuse_modules.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+        >>> m1 = nn.Conv2d(10, 20, 3)
+        >>> b1 = nn.BatchNorm2d(20)
+        >>> m2 = fuse_conv_bn(m1, b1)
+    """
+    assert(conv.training == bn.training),\
+        "Conv and BN both must be in the same mode (train or eval)."
+
+    if conv.training:
+        fused_conv = torch.nn._intrinsic.modules.ConvBn2d(conv, bn)
```
Comment (dzhulgakov) [inline torch/quantization/fuse_modules.py]: nit: more readable if you just put `return` here

File: test/test_quantization.py [hunk 2]
```diff
@@ -318,5 +320,55 @@ def checkQuantized(model):
...
+        self.assertEqual(type(testMod.sub1.conv), torch_fused.ConvBn2d,
+                         "Fused submodule Conv + BN")
+        self.assertEqual(type(testMod.sub1.bn), torch.nn.Identity,
+                         "Fused submodule Conv + BN (skipped BN)")
+        self.assertEqual(type(testMod.sub2.conv), torch.nn.Conv2d,
+                         "Non-fused submodule Conv")
+        self.assertEqual(type(testMod.sub2.bn), torch.nn.BatchNorm2d,
+                         "Non-fused submodule BN")
+
+    def test_fuse_module_eval(self):
```
Comment (dzhulgakov) [inline test/test_quantization.py]: do you want to add a test for numerics? specifically Conv+BN?

File: torch/quantization/fuse_modules.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        bn_mean = bn.running_mean
+        bn_var_sqrt = torch.sqrt(bn.running_var + bn.eps)
+
+        bn_weight = bn.weight
+        bn_bias = bn.bias
+
+        if b_conv is None:
+            b_conv = bn_mean.new_zeros(bn_mean.shape)
+
+        w_conv = w_conv * (bn_weight / bn_var_sqrt).reshape([-1, 1, 1, 1])
```
Comment (dzhulgakov) [inline torch/quantization/fuse_modules.py]: why are you reshaping it here?

File: torch/quantization/fuse_modules.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import torch
+import copy
+
+import torch.nn._intrinsic.modules.fused as torch_fused
+
+def fuse_conv_bn(conv, bn):
```
Comment (dzhulgakov) [inline torch/quantization/fuse_modules.py]: I feel I commented on one of the previous PRs, not sure which one. This is pretty general purpose - why it doesn't live in torch.nn.utils?

File: test/test_quantization.py (hunk 2 above)
Comment (raghuramank100) [reply test/test_quantization.py]: Please add tests for numerics for both eval and training cases for conv+bn, conv+bn+relu.

File: test/test_quantization.py [hunk 5]
```diff
@@ -318,5 +320,55 @@ def checkQuantized(model):
...
+        import torch.nn._intrinsic.modules.fused as torch_fused
+        testMod = ModForFusion()
+        testMod.eval()
+        fuse_modules(testMod, [['conv1', 'bn1', 'relu1'] ,
+                               ['sub1.conv', 'sub1.bn']])
+        self.assertEqual(type(testMod.conv1), torch_fused.ConvReLU2d,
+                         "Fused Conv + BN + Relu first layer (BN is folded)")
+        self.assertEqual(type(testMod.conv1[0]), torch.nn.Conv2d,
+                         "Fused Conv + BN + Relu (Conv + folded BN only)")
```
Comment (raghuramank100) [inline test/test_quantization.py]: NIT: Fix comment string

File: torch/quantization/fuse_modules.py (hunk 4 above)
Comment (raghuramank100) [reply torch/quantization/fuse_modules.py]: The fusion code for inference is general. The part for training is really needed only for quantization aware training. Is it ok to move this to nn.utils?

**Review by raghuramank100 — CHANGES_REQUESTED** [review]
Please see comments

**Comment by vadimkantorov** [comment]
@jerryzh168 A generic module ConvBn or a functional form (formula) to get fused weight / bias is still useful, even if the user has to call it manually

File: torch/quantization/fuse_modules.py (hunk 3 above)
Comment (zafartahirov) [reply torch/quantization/fuse_modules.py]: Because the w_conv and bn_weight have different dimensions: [*, 1, 1, 1] vs. [*,]. I couldn't do the broadcasting :/

Comment (dzhulgakov) [reply torch/quantization/fuse_modules.py]: Ah, sorry, I misread what precisely you're reshaping. It makes sense (unsqueeze would be another option, but this works too)

File: torch/quantization/fuse_modules.py (hunk 4 above)
Comment (zafartahirov) [reply torch/quantization/fuse_modules.py]: I don't mind either way. I guess, I'll move it to torch.nn.utils

**Comment by jerryzh168** [comment]
@vadimkantorov I suppose the reason to use ConvBn is for speed up? I think we'll delegate that to JIT. In eager mode we cannot cover all the use cases user might have, so we don't want to provide that as a general API.

consider a simple case where you might have conv and bn module defined, but in forward you might have one conv followed by bn, but another conv followed by something else, and in this case you cannot fuse these two modules.

File: test/test_quantization.py (hunk 2 above)
Comment (zafartahirov) [reply test/test_quantization.py]: I think testing for conv+bn is sufficient because +relu version of it reuses?

File: torch/quantization/fuse_modules.py (hunk 4 above)
Comment (zafartahirov) [reply torch/quantization/fuse_modules.py]: I am keeping the tests in the test_quantization, as there is no dedicated utils test file. LMK if I need to move it

Comment (jerryzh168) [reply torch/quantization/fuse_modules.py]: @dzhulgakov we had a discussion on the quantization working group, this is not going to cover all the cases, cc @ZolotukhinM

**Review by jerryzh168 — APPROVED** [review]

File: torch/quantization/fuse_modules.py (hunk 4 above)
Comment (zafartahirov) [reply torch/quantization/fuse_modules.py]: Tried moving it to the `nn.utils` -- breaks everything. @jerryzh168 mentioned, might be due to circular dependency. Can move it if we converge to that decision

**Comment by vadimkantorov** [comment]
@jerryzh168 in principle jit would solve this, but I think this fusion is really common (and it’s clear the fusion is possible and valid) and would simplify the life for users, until jit is stable

File: torch/quantization/fuse_modules.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+        >>> b1 = nn.BatchNorm2d(20)
+        >>> m2 = fuse_conv_bn(m1, b1)
+    """
+    assert(conv.training == bn.training),\
+        "Conv and BN both must be in the same mode (train or eval)."
+
+    if conv.training:
+        return torch.nn._intrinsic.ConvBn2d(conv, bn)
+    else:
+        fused_conv = copy.deepcopy(conv)
```
Comment (raghuramank100) [inline torch/quantization/fuse_modules.py]: Please re-factor the code from line 27 to 46 as a seperate function fuse_conv_bn_inference and move that code to nn/utils.

**Review by raghuramank100 — CHANGES_REQUESTED** [review]
Please move conv+bn fusion for inference to nn/utils

**Comment by zafartahirov** [comment]
> @jerryzh168 i agree, in principle jit would solve this, but I think this fusion is really common (and it’s clear the fusion is possible and valid) and would simplify the life for users, until jit is stable

I factored out the eval part of the fusion into `torch.nn.utils.fusion`. I hope that could be the first step towards building more fused ops.

**Comment by vadimkantorov** [comment]
[inplace ReLUDropout](https://github.com/pytorch/pytorch/issues/22109#issuecomment-504657961) is one, [inplace BatchNorm+ReLU](https://github.com/mapillary/inplace_abn) is another . Maybe if jit will implement these, they can be nice examples

File: torch/nn/utils/fusion.py [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    bn_mean = bn.running_mean
+    bn_var_sqrt = torch.sqrt(bn.running_var + bn.eps)
+
+    bn_weight = bn.weight
+    bn_bias = bn.bias
+
+    if b_conv is None:
+        b_conv = bn_mean.new_zeros(bn_mean.shape)
+
+    w_conv = w_conv * (bn_weight / bn_var_sqrt).reshape([-1, 1, 1, 1])
```
Comment (vadimkantorov) [inline torch/nn/utils/fusion.py]: two divisions by bn_var_sqrt could be replaced by multiplication if rsqrt is used instead of sqrt (not sure if it matters much)

Comment (zafartahirov) [reply torch/nn/utils/fusion.py]: Will send optimization as a separate diff to unblock

File: torch/quantization/fuse_modules.py (hunk 4 above)
Comment (zafartahirov) [reply torch/quantization/fuse_modules.py]: Moved a part of it :)

File: torch/nn/utils/fusion.py (hunk 7 above)
Comment (zafartahirov) [reply torch/nn/utils/fusion.py]: #23275
